Add bugpoint command

bjorn3 opened the pull request:
You should probably run this in release mode, as it is even more slow otherwise.

Fixes #532 

## Passes
* remove instruction
* replace instruction with single ret val with iconst
* replace instruction with trap
* remove ebb

Edit: added additional shrinking passes

bjorn3 commented:
This is ready for review.

Usage:

```sh
cargo run --release bugpoint crash.clif 2>/dev/null
████████████████████████████████████████████████████████████ pass 0 phase remove inst                 1124/1124 done
████████████████████████████████████████████████████████████ pass 0 phase replace inst with iconst     744/ 744 done
████████████████████████████████████████████████████████████ pass 0 phase replace inst with trap       744/ 744 done
████████████████████████████████████████████████████████████ pass 0 phase remove ebb                   121/ 121 done
████████████████████████████████████████████████████████████ pass 1 phase remove inst                  743/ 743 done
████████████████████████████████████████████████████████████ pass 1 phase replace inst with iconst     266/ 266 done
████████████████████████████████████████████████████████████ pass 1 phase replace inst with trap       266/ 266 done
████████████████████████████████████████████████████████████ pass 1 phase remove ebb                    87/  87 done
████████████████████████████████████████████████████████████ pass 2 phase remove inst                  263/ 263 done
████████████████████████████████████████████████████████████ pass 2 phase replace inst with iconst     220/ 220 done
████████████████████████████████████████████████████████████ pass 2 phase replace inst with trap       220/ 220 done
████████████████████████████████████████████████████████████ pass 2 phase remove ebb                    86/  86 done
████████████████████████████████████████████████████████████ pass 3 phase remove inst                  220/ 220 done
████████████████████████████████████████████████████████████ pass 3 phase replace inst with iconst     220/ 220 done
████████████████████████████████████████████████████████████ pass 3 phase replace inst with trap       220/ 220 done
████████████████████████████████████████████████████████████ pass 3 phase remove ebb                    86/  86 done

[... reduced function]

121 ebbs 1124 insts -> 86 ebbs 220 insts
```

bnjbvr commented on src/bugpoint.rs:
```diff
@@ -0,0 +1,312 @@
+    std::io::stdout().flush().unwrap(); // Flush stdout to sync with panic messages on stderr
+
+    match std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
+        if let Err(err) = cranelift_codegen::verifier::verify_function(&func, isa) {
```
I think there's a method to convert a Result into an Option directly.

bnjbvr commented on src/bugpoint.rs:
```diff
@@ -0,0 +1,312 @@
+    );
+}
+
+enum Res {
```
This name is a bit terse, how about CheckResult?

bnjbvr commented on src/bugpoint.rs:
```diff
@@ -0,0 +1,312 @@
+        Ok(Some(err)) => return Res::Verifier(err.to_string()),
+        Ok(None) => {}
+        Err(err) => {
+            // FIXME prevent verifier panic on removing ebb1
```
FIXME here.

bjorn3 replied:
The panic is: "EBB node must have one reachable predecessor".

If I run `cargo run compile` on the ir shown when inserting `println!("{}", func);` here, it gives a normal verifier error:

```
[...]
    jump ebb1
;   ^~~~~~~~~
; error: inst102: invalid ebb reference ebb1


ebb2:
    trap user0
[...]
```

bnjbvr replied:
I think there's a method to recompute successors and predecessors after an EBB has been introduced or removed, so using it might avoid this verifier issue.

bjorn3 replied:
I believe you are refering to [`Context::compute_cfg`](https://docs.rs/cranelift-codegen/0.30.0/cranelift_codegen/struct.Context.html#method.compute_cfg). Calling it wont solve it, as the `Context` is not reused.

nbp replied:
nit: One way would be to only reduce on panics which are keeping a similar error message. This way, even if the panic message changes, it would not be accepted as a reducing case.

bjorn3 replied:
I prefer to just fix the verifier. Having other panics being allowed would make it easier to find panics which you didnt even know existed. Maybe even make bugpoint reduce all different panics it found.

bnjbvr commented on src/bugpoint.rs:
```diff
@@ -0,0 +1,312 @@
+        reduce(isa, func, verbose);
+    }
+
+    //print!("{}", timing::take_current());
```
nit: comment

bnjbvr commented on src/bugpoint.rs:
```diff
@@ -0,0 +1,312 @@
+            }
+            Phase::ReplaceInstWithTrap(ref mut ebb, ref mut inst) => {
+                if let Some((_prev_ebb, prev_inst)) = next_inst_ret_prev(func, ebb, inst) {
+                    func.dfg.replace(prev_inst).trap(TrapCode::User(0));
```
Replacing an instruction by a trap (which is a terminal instruction) in the middle of a block means that this block has to be split, which is likely to reduce the number of verifier errors encountered while reducing. Maybe a good idea for a follow-up!

bnjbvr commented on src/bugpoint.rs:
```diff
@@ -0,0 +1,312 @@
+            match check_for_crash(isa, &func2) {
+                Res::Succeed => {
+                    // Shrinking didn't hit the problem anymore, discard changes.
+                    //progress.println("succeeded");
```
nit: here and below, commented calls to progress

bnjbvr commented on src/bugpoint.rs:
```diff
@@ -0,0 +1,312 @@
+//! CLI tool to read Cranelift IR files and compile them into native code.
```
nit: this comment needs an update

bnjbvr commented on src/bugpoint.rs:
```diff
@@ -0,0 +1,312 @@
+    let fisa = parsed.as_fisa();
+
+    let path = Path::new(&filename).to_path_buf();
+    let name = String::from(path.as_os_str().to_string_lossy());
```
Why is the round trip from filename to name needed, here?

bjorn3 replied:
It isnt I think. I copy pasted and inlined some functions from elsewhere.

bnjbvr commented on src/bugpoint.rs:
```diff
@@ -0,0 +1,312 @@
+}
+
+fn reduce(isa: &TargetIsa, mut func: Function, verbose: bool) {
+    let (orig_ebb_count, orig_inst_count) = (ebb_count(&func), inst_count(&func));
```
We probably don't want to reduce a function which compilation doesn't panic in the first place, right?

bjorn3 replied:
Yes

bnjbvr commented on src/bugpoint.rs:
```diff
@@ -0,0 +1,312 @@
+    Panic,
+}
+
+fn check_for_crash(isa: &TargetIsa, func: &Function) -> Res {
```
There can be several kinds of panics in the pipeline, and I'm afraid that checking for a crash as is could hide the initial one we're looking for, and instead cause a panic in an unrelated location. It would be nice to handle this case, e.g. by checking that the panic reason is the same as the one we observed during the first run I'm talking about above. Bonus points if, in this case, we can still have the clif causing the new panic.

bjorn3 commented:
I will make the requested changes in a few days.

bjorn3 commented:
@bnjbvr I have some more reduction methods in mind (eg replacing ebb params with `iconst` (`jump ebb1(v1) ebb1(v1: i8): ...` to `jump ebb1 ebb1: v1 = iconst.i8 0 ...`) Do you want me to open a new issue to track them or do you want to keep #532 open?

bnjbvr commented:
> @bnjbvr I have some more reduction methods in mind (eg replacing ebb params with iconst (jump ebb1(v1) ebb1(v1: i8): ... to jump ebb1 ebb1: v1 = iconst.i8 0 ...)) Do you want me to open a new issue to track them or do you want to keep #532 open?

New issue sounds fine, thanks.

bjorn3 commented:
Ping

bnjbvr commented:
I'll try to review this this week.

bnjbvr commented:
@nbp said he could help with this review. Thanks!

nbp commented:
I will look at this issue tomorrow.

bjorn3 commented on src/bugpoint.rs:
```diff
@@ -0,0 +1,320 @@
+}
+
+impl Phase {
+    fn step(&mut self, func: &mut Function) -> PhaseStepResult {
```
Is it better now?

nbp replied:
Yes!

nbp commented on src/bugpoint.rs:
```diff
@@ -0,0 +1,423 @@
+}
+
+impl RemoveInst {
+    fn new(func: &Function) -> Box<dyn Mutator> {
```
nit: This is an unexpected signature for `new` function in Rust. Maybe move the boxing outside of this function.

nbp commented on src/bugpoint.rs:
```diff
@@ -0,0 +1,423 @@
+        for _ in 0..10000 {
+            progress.inc(1);
+
+            let (mutated_func, msg, mutation_kind) = match self.mutate(func.clone()) {
```
Note: There is one thing which conceptually slightly troubles me a bit about this code. Here we do a `func.clone()`, but a mutator such as `RemoveInst` keep both an instruction and block index.  Technically these indexes should only be valid within a given function, and using them in a clone is abusing the entity system of Cranelift for not keeping track of its primary and secondary vectors with which they can safely be used.

While I admit that this works in practice, this could be a pain point if we were to make a more secure entity system.

bjorn3 replied:
It has to clone anyway, to be able to revert mutations that are not resulting in the wanted error.

nbp commented on src/bugpoint.rs:
```diff
@@ -0,0 +1,423 @@
+
+    resolve_aliases(&mut func);
+
+    'outer_loop: for pass_idx in 0..100 {
```
nit: remove the `outer_loop` label.

bjorn3 commented:
I added a pass to remove unused entities (stack slots, ext funcs, global values). I also added a test to check that bugpoint does actually succeed in reducing a test case.
